File: Golang/Interview_problems/TradingSystem_LLD.go
/*
================================================================================
LLD: Trading System (Stock Exchange Order Matching Engine)
================================================================================

Functional Requirements:
    1. Place a LIMIT order (BUY or SELL) for a symbol with a price and
       quantity.
    2. Match incoming orders against resting orders in the order book using
       price-time priority: best price first, earliest order first at a
       given price.
    3. Support partial fills: an order can match against several resting
       orders, and a resting order can be partially filled and stay in the
       book for future matches.
    4. Cancel a resting (OPEN or PARTIALLY_FILLED) order.
    5. Query an order's current status and the full trade history.

Non-Functional Requirements:
    1. Thread-safety.
    2. Maintainability / extensibility to new order types and matching
       strategies.
    3. Performance: matching should only look at the best resting orders,
       not rescan the whole book for every order.

Design:
    MatchingStrategy (Strategy) implements the matching algorithm;
    PriceTimePriorityMatching walks the opposite side of the book from the
    best price outward (ties broken by earliest timestamp), filling the
    incoming order against one or more resting orders at each resting
    order's own price, until the incoming order is filled or no resting
    order crosses its price.

    OrderBook holds the resting BUY and SELL orders for one symbol and
    exposes best-price lookups, insertion, and removal.

    MatchingEngine owns one OrderBook per symbol and delegates matching to
    the MatchingStrategy; PlaceOrder() runs the match and rests whatever
    quantity is left, CancelOrder() removes a resting order from its book.

    TradingSystem (facade) wires the engine together with order/trade
    storage: PlaceOrder() creates the order, runs it through the engine,
    and records any resulting trades; CancelOrder(), GetOrderStatus() and
    GetTradeHistory() serve the remaining queries.

Core Entities:
    Side, OrderType, OrderStatus (enums)
    Order
    Trade
    OrderBook
    MatchingStrategy / PriceTimePriorityMatching
    MatchingEngine
    TradingSystem
================================================================================
*/

package main

import (
	"errors"
	"fmt"
	"sync"
)

type Side int

const (
	Buy Side = iota
	Sell
)

func (s Side) String() string {
	if s == Buy {
		return "BUY"
	}
	return "SELL"
}

type OrderType int

const (
	Limit OrderType = iota
)

func (t OrderType) String() string {
	return "LIMIT"
}

type OrderStatus int

const (
	Open OrderStatus = iota
	PartiallyFilled
	Filled
	Cancelled
)

func (s OrderStatus) String() string {
	switch s {
	case Open:
		return "OPEN"
	case PartiallyFilled:
		return "PARTIALLY_FILLED"
	case Filled:
		return "FILLED"
	case Cancelled:
		return "CANCELLED"
	}
	return "UNKNOWN"
}

type Order struct {
	ID                string
	Symbol            string
	Side              Side
	Type              OrderType
	Price             float64
	Quantity          int
	RemainingQuantity int
	Status            OrderStatus
	Timestamp         int64
}

type Trade struct {
	ID          string
	BuyOrderID  string
	SellOrderID string
	Symbol      string
	Price       float64
	Quantity    int
}

type OrderBook struct {
	Symbol     string
	BuyOrders  []*Order
	SellOrders []*Order
}

func NewOrderBook(symbol string) *OrderBook {
	return &OrderBook{Symbol: symbol}
}

func (b *OrderBook) AddOrder(order *Order) {
	if order.Side == Buy {
		b.BuyOrders = append(b.BuyOrders, order)
	} else {
		b.SellOrders = append(b.SellOrders, order)
	}
}

func (b *OrderBook) BestBuy() *Order {
	var best *Order
	for _, o := range b.BuyOrders {
		if best == nil || o.Price > best.Price || (o.Price == best.Price && o.Timestamp < best.Timestamp) {
			best = o
		}
	}
	return best
}

func (b *OrderBook) BestSell() *Order {
	var best *Order
	for _, o := range b.SellOrders {
		if best == nil || o.Price < best.Price || (o.Price == best.Price && o.Timestamp < best.Timestamp) {
			best = o
		}
	}
	return best
}

func (b *OrderBook) RemoveOrder(order *Order) bool {
	list := &b.BuyOrders
	if order.Side == Sell {
		list = &b.SellOrders
	}
	for i, o := range *list {
		if o == order {
			*list = append((*list)[:i], (*list)[i+1:]...)
			return true
		}
	}
	return false
}

type MatchingStrategy interface {
	Match(book *OrderBook, incoming *Order) []Trade
}

type PriceTimePriorityMatching struct {
	tradeSequence *int64
}

func (m PriceTimePriorityMatching) Match(book *OrderBook, incoming *Order) []Trade {
	var trades []Trade

	for incoming.RemainingQuantity > 0 {
		var resting *Order
		if incoming.Side == Buy {
			resting = book.BestSell()
		} else {
			resting = book.BestBuy()
		}
		if resting == nil {
			break
		}
		if incoming.Side == Buy && resting.Price > incoming.Price {
			break
		}
		if incoming.Side == Sell && resting.Price < incoming.Price {
			break
		}

		tradeQty := incoming.RemainingQuantity
		if resting.RemainingQuantity < tradeQty {
			tradeQty = resting.RemainingQuantity
		}
		tradePrice := resting.Price
		buyOrderID, sellOrderID := incoming.ID, resting.ID
		if incoming.Side == Sell {
			buyOrderID, sellOrderID = resting.ID, incoming.ID
		}
		*m.tradeSequence++
		trades = append(trades, Trade{
			ID: fmt.Sprintf("T%d", *m.tradeSequence), BuyOrderID: buyOrderID, SellOrderID: sellOrderID,
			Symbol: book.Symbol, Price: tradePrice, Quantity: tradeQty,
		})

		incoming.RemainingQuantity -= tradeQty
		resting.RemainingQuantity -= tradeQty

		if resting.RemainingQuantity == 0 {
			resting.Status = Filled
			book.RemoveOrder(resting)
		} else {
			resting.Status = PartiallyFilled
		}
	}

	if incoming.RemainingQuantity == 0 {
		incoming.Status = Filled
	} else {
		if incoming.RemainingQuantity < incoming.Quantity {
			incoming.Status = PartiallyFilled
		} else {
			incoming.Status = Open
		}
		book.AddOrder(incoming)
	}
	return trades
}

var (
	ErrOrderNotFound       = errors.New("unknown order")
	ErrOrderNotCancellable = errors.New("order cannot be cancelled")
)

type MatchingEngine struct {
	matchingStrategy MatchingStrategy
	books            map[string]*OrderBook
}

func NewMatchingEngine(matchingStrategy MatchingStrategy) *MatchingEngine {
	return &MatchingEngine{matchingStrategy: matchingStrategy, books: make(map[string]*OrderBook)}
}

func (e *MatchingEngine) bookFor(symbol string) *OrderBook {
	book, ok := e.books[symbol]
	if !ok {
		book = NewOrderBook(symbol)
		e.books[symbol] = book
	}
	return book
}

func (e *MatchingEngine) PlaceOrder(order *Order) []Trade {
	return e.matchingStrategy.Match(e.bookFor(order.Symbol), order)
}

func (e *MatchingEngine) CancelOrder(order *Order) bool {
	book, ok := e.books[order.Symbol]
	if !ok {
		return false
	}
	removed := book.RemoveOrder(order)
	if removed {
		order.Status = Cancelled
	}
	return removed
}

func (e *MatchingEngine) BookForSymbol(symbol string) *OrderBook {
	return e.books[symbol]
}

type OrderPlacementResult struct {
	Order  *Order
	Trades []Trade
}

type TradingSystem struct {
	mu             sync.Mutex
	matchingEngine *MatchingEngine
	ordersByID     map[string]*Order
	tradeHistory   []Trade
	orderSequence  int64
}

func NewTradingSystem(matchingEngine *MatchingEngine) *TradingSystem {
	return &TradingSystem{matchingEngine: matchingEngine, ordersByID: make(map[string]*Order)}
}

func (s *TradingSystem) PlaceOrder(symbol string, side Side, orderType OrderType, price float64, quantity int) OrderPlacementResult {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.orderSequence++
	order := &Order{
		ID: fmt.Sprintf("O%d", s.orderSequence), Symbol: symbol, Side: side, Type: orderType,
		Price: price, Quantity: quantity, RemainingQuantity: quantity, Status: Open, Timestamp: s.orderSequence,
	}
	s.ordersByID[order.ID] = order
	trades := s.matchingEngine.PlaceOrder(order)
	s.tradeHistory = append(s.tradeHistory, trades...)
	return OrderPlacementResult{Order: order, Trades: trades}
}

func (s *TradingSystem) CancelOrder(orderID string) (*Order, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	order, ok := s.ordersByID[orderID]
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrOrderNotFound, orderID)
	}
	if order.Status != Open && order.Status != PartiallyFilled {
		return nil, fmt.Errorf("%w: order %s is %s", ErrOrderNotCancellable, orderID, order.Status)
	}
	s.matchingEngine.CancelOrder(order)
	return order, nil
}

func (s *TradingSystem) GetOrderStatus(orderID string) (OrderStatus, bool) {
	order, ok := s.ordersByID[orderID]
	if !ok {
		return 0, false
	}
	return order.Status, true
}

func (s *TradingSystem) GetTradeHistory() []Trade {
	s.mu.Lock()
	defer s.mu.Unlock()
	history := make([]Trade, len(s.tradeHistory))
	copy(history, s.tradeHistory)
	return history
}

func (s *TradingSystem) GetRestingBuyOrders(symbol string) []*Order {
	book := s.matchingEngine.BookForSymbol(symbol)
	if book == nil {
		return nil
	}
	return book.BuyOrders
}

func (s *TradingSystem) GetRestingSellOrders(symbol string) []*Order {
	book := s.matchingEngine.BookForSymbol(symbol)
	if book == nil {
		return nil
	}
	return book.SellOrders
}

func describePlacement(result OrderPlacementResult) {
	order := result.Order
	fmt.Printf("  %s: %s %d @ %.2f\n", order.ID, order.Side, order.Quantity, order.Price)
	if len(result.Trades) == 0 {
		fmt.Printf("    no match, resting in book (status %s)\n", order.Status)
		return
	}
	for _, t := range result.Trades {
		fmt.Printf("    matched %d @ %.2f (trade %s)\n", t.Quantity, t.Price, t.ID)
	}
	remaining := ""
	if order.Status == PartiallyFilled {
		remaining = fmt.Sprintf(" (remaining %d)", order.RemainingQuantity)
	}
	fmt.Printf("    %s status now: %s%s\n", order.ID, order.Status, remaining)
}

func main() {
	var tradeSequence int64
	system := NewTradingSystem(NewMatchingEngine(PriceTimePriorityMatching{tradeSequence: &tradeSequence}))

	fmt.Println("Placing orders on AAPL:")

	s1 := system.PlaceOrder("AAPL", Sell, Limit, 101.00, 10)
	describePlacement(s1)

	s2 := system.PlaceOrder("AAPL", Sell, Limit, 100.50, 5)
	describePlacement(s2)

	s3 := system.PlaceOrder("AAPL", Sell, Limit, 100.75, 3)
	describePlacement(s3)

	fmt.Println("\nA sweeping buy order that crosses multiple price levels:")
	b1 := system.PlaceOrder("AAPL", Buy, Limit, 101.00, 12)
	describePlacement(b1)

	fmt.Println("\nA buy order priced too low to match anything:")
	b2 := system.PlaceOrder("AAPL", Buy, Limit, 99.00, 2)
	describePlacement(b2)

	fmt.Printf("\nCancelling %s while it still rests in the book:\n", b2.Order.ID)
	cancelled, _ := system.CancelOrder(b2.Order.ID)
	fmt.Printf("  %s status now: %s\n", cancelled.ID, cancelled.Status)

	fmt.Println("\nFinal order book for AAPL:")
	fmt.Println("  BUY side:")
	for _, o := range system.GetRestingBuyOrders("AAPL") {
		fmt.Printf("    %s: %d @ %.2f (remaining %d, %s)\n", o.ID, o.Quantity, o.Price, o.RemainingQuantity, o.Status)
	}
	fmt.Println("  SELL side:")
	for _, o := range system.GetRestingSellOrders("AAPL") {
		fmt.Printf("    %s: %d @ %.2f (remaining %d, %s)\n", o.ID, o.Quantity, o.Price, o.RemainingQuantity, o.Status)
	}

	fmt.Println("\nTrade history:")
	for _, t := range system.GetTradeHistory() {
		fmt.Printf("    %s: %s bought from %s, %d @ %.2f\n", t.ID, t.BuyOrderID, t.SellOrderID, t.Quantity, t.Price)
	}
}

File: Java/Interview_problems/TradingSystem_LLD.java
/*
================================================================================
LLD: Trading System (Stock Exchange Order Matching Engine)
================================================================================

Functional Requirements:
    1. Place a LIMIT order (BUY or SELL) for a symbol with a price and
       quantity.
    2. Match incoming orders against resting orders in the order book using
       price-time priority: best price first, earliest order first at a
       given price.
    3. Support partial fills: an order can match against several resting
       orders, and a resting order can be partially filled and stay in the
       book for future matches.
    4. Cancel a resting (OPEN or PARTIALLY_FILLED) order.
    5. Query an order's current status and the full trade history.

Non-Functional Requirements:
    1. Thread-safety.
    2. Maintainability / extensibility to new order types and matching
       strategies.
    3. Performance: matching should only look at the best resting orders,
       not rescan the whole book for every order.

Design:
    MatchingStrategy (Strategy) implements the matching algorithm;
    PriceTimePriorityMatching walks the opposite side of the book from the
    best price outward (ties broken by earliest timestamp), filling the
    incoming order against one or more resting orders at each resting
    order's own price, until the incoming order is filled or no resting
    order crosses its price.

    OrderBook holds the resting BUY and SELL orders for one symbol and
    exposes best-price lookups, insertion, and removal.

    MatchingEngine owns one OrderBook per symbol and delegates matching to
    the MatchingStrategy; placeOrder() runs the match and rests whatever
    quantity is left, cancelOrder() removes a resting order from its book.

    TradingSystem (facade) wires the engine together with order/trade
    storage: placeOrder() creates the order, runs it through the engine,
    and records any resulting trades; cancelOrder(), getOrderStatus() and
    getTradeHistory() serve the remaining queries.

Core Entities:
    Side, OrderType, OrderStatus (enums)
    Order
    Trade
    OrderBook
    MatchingStrategy / PriceTimePriorityMatching
    MatchingEngine
    TradingSystem
================================================================================
*/

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicLong;

enum Side {
    BUY,
    SELL
}

enum OrderType {
    LIMIT
}

enum OrderStatus {
    OPEN,
    PARTIALLY_FILLED,
    FILLED,
    CANCELLED
}

class Order {
    private static final AtomicLong SEQUENCE = new AtomicLong(0);

    final String id;
    final String symbol;
    final Side side;
    final OrderType type;
    final double price;
    final int quantity;
    int remainingQuantity;
    OrderStatus status = OrderStatus.OPEN;
    final long timestamp;

    Order(String id, String symbol, Side side, OrderType type, double price, int quantity) {
        this.id = id;
        this.symbol = symbol;
        this.side = side;
        this.type = type;
        this.price = price;
        this.quantity = quantity;
        this.remainingQuantity = quantity;
        this.timestamp = SEQUENCE.incrementAndGet();
    }
}

class Trade {
    private static final AtomicLong SEQUENCE = new AtomicLong(0);

    final String id;
    final String buyOrderId;
    final String sellOrderId;
    final String symbol;
    final double price;
    final int quantity;

    Trade(String buyOrderId, String sellOrderId, String symbol, double price, int quantity) {
        this.id = "T" + SEQUENCE.incrementAndGet();
        this.buyOrderId = buyOrderId;
        this.sellOrderId = sellOrderId;
        this.symbol = symbol;
        this.price = price;
        this.quantity = quantity;
    }
}

class OrderBook {
    final String symbol;
    final List<Order> buyOrders = new ArrayList<>();
    final List<Order> sellOrders = new ArrayList<>();

    OrderBook(String symbol) {
        this.symbol = symbol;
    }

    void addOrder(Order order) {
        if (order.side == Side.BUY) buyOrders.add(order);
        else sellOrders.add(order);
    }

    Order bestBuy() {
        Order best = null;
        for (Order o : buyOrders) {
            if (best == null || o.price > best.price || (o.price == best.price && o.timestamp < best.timestamp)) {
                best = o;
            }
        }
        return best;
    }

    Order bestSell() {
        Order best = null;
        for (Order o : sellOrders) {
            if (best == null || o.price < best.price || (o.price == best.price && o.timestamp < best.timestamp)) {
                best = o;
            }
        }
        return best;
    }

    boolean removeOrder(Order order) {
        if (order.side == Side.BUY) return buyOrders.remove(order);
        return sellOrders.remove(order);
    }
}

interface MatchingStrategy {
    List<Trade> match(OrderBook book, Order incoming);
}

class PriceTimePriorityMatching implements MatchingStrategy {
    @Override
    public List<Trade> match(OrderBook book, Order incoming) {
        List<Trade> trades = new ArrayList<>();

        while (incoming.remainingQuantity > 0) {
            Order resting = incoming.side == Side.BUY ? book.bestSell() : book.bestBuy();
            if (resting == null) break;
            if (incoming.side == Side.BUY && resting.price > incoming.price) break;
            if (incoming.side == Side.SELL && resting.price < incoming.price) break;

            int tradeQty = Math.min(incoming.remainingQuantity, resting.remainingQuantity);
            double tradePrice = resting.price;
            String buyOrderId = incoming.side == Side.BUY ? incoming.id : resting.id;
            String sellOrderId = incoming.side == Side.BUY ? resting.id : incoming.id;
            trades.add(new Trade(buyOrderId, sellOrderId, book.symbol, tradePrice, tradeQty));

            incoming.remainingQuantity -= tradeQty;
            resting.remainingQuantity -= tradeQty;

            if (resting.remainingQuantity == 0) {
                resting.status = OrderStatus.FILLED;
                book.removeOrder(resting);
            } else {
                resting.status = OrderStatus.PARTIALLY_FILLED;
            }
        }

        if (incoming.remainingQuantity == 0) {
            incoming.status = OrderStatus.FILLED;
        } else {
            incoming.status = incoming.remainingQuantity < incoming.quantity
                    ? OrderStatus.PARTIALLY_FILLED
                    : OrderStatus.OPEN;
            book.addOrder(incoming);
        }
        return trades;
    }
}

class OrderNotFoundException extends RuntimeException {
    OrderNotFoundException(String message) {
        super(message);
    }
}

class OrderNotCancellableException extends RuntimeException {
    OrderNotCancellableException(String message) {
        super(message);
    }
}

class MatchingEngine {
    private final Map<String, OrderBook> books = new LinkedHashMap<>();
    private final MatchingStrategy matchingStrategy;

    MatchingEngine(MatchingStrategy matchingStrategy) {
        this.matchingStrategy = matchingStrategy;
    }

    private OrderBook bookFor(String symbol) {
        return books.computeIfAbsent(symbol, OrderBook::new);
    }

    List<Trade> placeOrder(Order order) {
        return matchingStrategy.match(bookFor(order.symbol), order);
    }

    boolean cancelOrder(Order order) {
        OrderBook book = books.get(order.symbol);
        if (book == null) return false;
        boolean removed = book.removeOrder(order);
        if (removed) order.status = OrderStatus.CANCELLED;
        return removed;
    }

    OrderBook bookForSymbol(String symbol) {
        return books.get(symbol);
    }
}

class OrderPlacementResult {
    final Order order;
    final List<Trade> trades;

    OrderPlacementResult(Order order, List<Trade> trades) {
        this.order = order;
        this.trades = trades;
    }
}

class TradingSystem {
    private final MatchingEngine matchingEngine;
    private final Map<String, Order> ordersById = new LinkedHashMap<>();
    private final List<Trade> tradeHistory = new ArrayList<>();
    private final AtomicLong orderSequence = new AtomicLong(0);
    private final Object lock = new Object();

    TradingSystem(MatchingEngine matchingEngine) {
        this.matchingEngine = matchingEngine;
    }

    OrderPlacementResult placeOrder(String symbol, Side side, OrderType type, double price, int quantity) {
        synchronized (lock) {
            String orderId = "O" + orderSequence.incrementAndGet();
            Order order = new Order(orderId, symbol, side, type, price, quantity);
            ordersById.put(orderId, order);
            List<Trade> trades = matchingEngine.placeOrder(order);
            tradeHistory.addAll(trades);
            return new OrderPlacementResult(order, trades);
        }
    }

    Order cancelOrder(String orderId) {
        synchronized (lock) {
            Order order = ordersById.get(orderId);
            if (order == null) {
                throw new OrderNotFoundException("Unknown order " + orderId);
            }
            if (order.status != OrderStatus.OPEN && order.status != OrderStatus.PARTIALLY_FILLED) {
                throw new OrderNotCancellableException(
                        "Order " + orderId + " is " + order.status + " and cannot be cancelled");
            }
            matchingEngine.cancelOrder(order);
            return order;
        }
    }

    Optional<OrderStatus> getOrderStatus(String orderId) {
        Order order = ordersById.get(orderId);
        return order == null ? Optional.empty() : Optional.of(order.status);
    }

    List<Trade> getTradeHistory() {
        synchronized (lock) {
            return new ArrayList<>(tradeHistory);
        }
    }

    List<Order> getRestingBuyOrders(String symbol) {
        OrderBook book = matchingEngine.bookForSymbol(symbol);
        return book == null ? List.of() : new ArrayList<>(book.buyOrders);
    }

    List<Order> getRestingSellOrders(String symbol) {
        OrderBook book = matchingEngine.bookForSymbol(symbol);
        return book == null ? List.of() : new ArrayList<>(book.sellOrders);
    }
}

public class TradingSystem_LLD {
    public static void main(String[] args) {
        TradingSystem system = new TradingSystem(new MatchingEngine(new PriceTimePriorityMatching()));

        System.out.println("Placing orders on AAPL:");

        OrderPlacementResult s1 = system.placeOrder("AAPL", Side.SELL, OrderType.LIMIT, 101.00, 10);
        describePlacement(s1);

        OrderPlacementResult s2 = system.placeOrder("AAPL", Side.SELL, OrderType.LIMIT, 100.50, 5);
        describePlacement(s2);

        OrderPlacementResult s3 = system.placeOrder("AAPL", Side.SELL, OrderType.LIMIT, 100.75, 3);
        describePlacement(s3);

        System.out.println("\nA sweeping buy order that crosses multiple price levels:");
        OrderPlacementResult b1 = system.placeOrder("AAPL", Side.BUY, OrderType.LIMIT, 101.00, 12);
        describePlacement(b1);

        System.out.println("\nA buy order priced too low to match anything:");
        OrderPlacementResult b2 = system.placeOrder("AAPL", Side.BUY, OrderType.LIMIT, 99.00, 2);
        describePlacement(b2);

        System.out.println("\nCancelling " + b2.order.id + " while it still rests in the book:");
        Order cancelled = system.cancelOrder(b2.order.id);
        System.out.println("  " + cancelled.id + " status now: " + cancelled.status);

        System.out.println("\nFinal order book for AAPL:");
        System.out.println("  BUY side:");
        for (Order o : system.getRestingBuyOrders("AAPL")) {
            System.out.printf("    %s: %d @ %.2f (remaining %d, %s)%n", o.id, o.quantity, o.price, o.remainingQuantity, o.status);
        }
        System.out.println("  SELL side:");
        for (Order o : system.getRestingSellOrders("AAPL")) {
            System.out.printf("    %s: %d @ %.2f (remaining %d, %s)%n", o.id, o.quantity, o.price, o.remainingQuantity, o.status);
        }

        System.out.println("\nTrade history:");
        for (Trade t : system.getTradeHistory()) {
            System.out.printf("    %s: %s bought from %s, %d @ %.2f%n", t.id, t.buyOrderId, t.sellOrderId, t.quantity, t.price);
        }
    }

    private static void describePlacement(OrderPlacementResult result) {
        Order order = result.order;
        System.out.printf("  %s: %s %d @ %.2f%n", order.id, order.side, order.quantity, order.price);
        if (result.trades.isEmpty()) {
            System.out.println("    no match, resting in book (status " + order.status + ")");
        } else {
            for (Trade t : result.trades) {
                System.out.printf("    matched %d @ %.2f (trade %s)%n", t.quantity, t.price, t.id);
            }
            System.out.println("    " + order.id + " status now: " + order.status
                    + (order.status == OrderStatus.PARTIALLY_FILLED ? " (remaining " + order.remainingQuantity + ")" : ""));
        }
    }
}
